### ISSUE TITLE
RFC: Remove item-level updates from IGListCollectionContext, require them done in batch updates

### ISSUE THREAD

**Issue by rnystrom** [Instagram/IGListKit#392]
It is way too easy to execute an item-level update out-of-sync with coalesced batch updates and break things. We've observed this in Instagram. Although rare, at our scale its still a measurable problem.

By item-level updates, I'm specifically talking about:

- [Reload items](https://github.com/Instagram/IGListKit/blob/master/Source/IGListCollectionContext.h#L176-L183)
- [Insert items](https://github.com/Instagram/IGListKit/blob/master/Source/IGListCollectionContext.h#L185-L192)
- [Delete items](https://github.com/Instagram/IGListKit/blob/master/Source/IGListCollectionContext.h#L194-L201)
- [Reload](https://github.com/Instagram/IGListKit/blob/master/Source/IGListCollectionContext.h#L203-L208) the whole section controller

## The Problem

The preferred way to use them is to [batch update](https://github.com/Instagram/IGListKit/blob/master/Source/IGListCollectionContext.h#L210-L240) your **state** (e.g. w/e powers `numberOfItems`) and **item** changes:

```objc
[self.collectionContext performBatchAnimated:YES updates:^{
  NSMutableArray *models = [self.models mutableCopy];
  [models removeObjectAtIndex:2];
  self.models = models;
  [self.collectionContext deleteInSectionController:self 
    atIndexes:[NSIndexSet indexSetWithIndex:2]];
} completion: nil]
```

However I've seen lots of examples not-using the batch update block:

```objc
NSMutableArray *models = [self.models mutableCopy];
[models removeObjectAtIndex:2];
self.models = models;
[self.collectionContext deleteInSectionController:self 
  atIndexes:[NSIndexSet indexSetWithIndex:2]];
```

We even do this in one of [the examples](https://github.com/Instagram/IGListKit/blob/master/Examples/Examples-iOS/IGListKitExamples/SectionControllers/ExpandableSectionController.swift#L45-L46).

The problem w/ this is that if a batch update is in progress, you might update while the `UICollectionViewData` (private structure) is changing the state of the `UICollectionView`. While this is all main-thread affined, the whole operation happens over multiple runloop turns, so it should be treated as an async operation (e.g. racing).

## Proposal

I propose we remove all of the item-update operations from `IGListCollectionContext` so they cannot be done w/out batch updates. Instead, we provide a new `IGListBatchUpdateContext` object as a parameter to the update block that has all of the above batch update methods.

So now your batch updates would look like:

```objc
[self.collectionContext performBatchAnimated:YES updates:^(IGListBatchUpdateContext *context) {
  NSMutableArray *models = [self.models mutableCopy];
  [models removeObjectAtIndex:2];
  self.models = models;
  [context deleteInSectionController:self atIndexes:[NSIndexSet indexSetWithIndex:2]];
} completion: nil]
```

## Considerations

That way the design and compiler forces when we can update. The only thing we _can't_ enforce is when the **state changes** happen. We can use documentation and examples to try get the message across, but there's nothing stopping someone from doing:

```objc
// BAD!
NSMutableArray *models = [self.models mutableCopy];
[models removeObjectAtIndex:2];
self.models = models;

[self.collectionContext performBatchAnimated:YES updates:^(IGListBatchUpdateContext *context) {
  [context deleteInSectionController:self atIndexes:[NSIndexSet indexSetWithIndex:2]];
} completion: nil]
```

Maybe we could add some fancy asserting or something just before [performBatchUpdates:](https://github.com/Instagram/IGListKit/blob/master/Source/IGListAdapterUpdater.m#L212) that checks the data source's `numberOfSections` and `numberOfItemsInSection:` and throws if the data source and UICV are out of sync? 

That might be useful as its own thing. We could pinpoint which section controller is borked.

**Comment by jessesquires** [comment]
💯

**Comment by Adlai-Holler** [comment]
Wow this is great! I just started learning IGListKit a couple days ago and was wondering how those two update APIs would play together. 

What would it mean for two section controllers to do something like this in the same run loop? Would the updates be coalesced and performed all in one batch?

Tangential but related: if I wanted to build a fully data driven arch (~ #38) would something like this be appropriate?

```objc
// Pseudocode
- (void)didUpdateToObject:(id)object {
  self.model = object;
  NSArray *oldItems = self.items;
  self.items = [self.class itemsFromObject:object];
  IGListDiff *diff = …;
  [self.collectionContext applyDiff:diff inSectionController:self];
}
```

Another tangential point, would `performUpdatesAnimated:additionalUpdates:completion:` or something similar be more appropriate? It (1) shows that this is the method for updating whether it's batch or no, and (2) indicates that the updates you provide are in addition to the section-level updates from the list adapter.

**Comment by heumn** [comment]
> Maybe we could add some fancy asserting or something just before performBatchUpdates: that checks the data source's numberOfSections and numberOfItemsInSection: and throws if the data source and UICV are out of sync?

Here the consumer of the API is alredy on the path down a slipperly slope of "unkowns errors so if you can help him "do the right thing" early on, this sounds like a good idea.

**Comment by rnystrom** [comment]
Diff up internally to make this change! 🎉

**Comment by broadwaylamb** [comment]
This change actually broke my project. I had a list of sections where each section could be expanded when selected, but if I selected another section, the previously selected section would collapse and then the new section would expand. By expanding a section I mean adding some cells to it.

In a section controller I maintained an `expanded` variable and when it changed its value I would call `reloadSectionController`. So basically there were two `reloadSectionController` calls at a time (one for collapsing the previously expanded section and the other for expanding the new one) and it worked fine. But now it doesn't.

Calling `- performBatchAnimated:updates:completion` twice at a time causes an exception.


Here is the code I used before:

```swift
var expanded = false {
    didSet {
        collectionContext?.reload(self)
    }
}
```

and after:
```swift
var expanded = false {
    didSet {

        let updates = { (context: IGListBatchContext) in
            context.reload(self)
        }

        collectionContext?
            .performBatch(animated: true, updates: updates, completion: nil)
    }
}
```
